docs/rules/no-test-callback.md

```diff
@@ -0,0 +1,76 @@
+# Avoid using a callback in asynchronous tests (no-test-callback)
+
+Jest allows you to pass a callback to test definitions, typically called `done`,
+that is later invoked to indicate that the asynchronous test is complete.
+
+However, that means that if your test throws (e.g. because of a failing
+assertion), `done` will never be called unless you manually use `try-catch`.
+
+```js
+test('some test', done => {
+  expect(false).toBe(true);
+  done();
+});
+```
+
+The test above will time out instead of failing the assertions, since `done` is
+never called.
+
+Correct way of doing the same thing is to wrap it in `try-catch`.
+
+```js
+test('some test', done => {
+  try {
+    expect(false).toBe(true);
+    done();
+  } catch (e) {
+    done(e);
+  }
+});
+```
+
+However, Jest supports a second way of having asynchronous tests - using
+promises.
+
+```js
+test('some test', () => {
+  return new Promise(done => {
+    expect(false).toBe(true);
+    done();
```

Comment:
I think wrapping this code in a Promise as a fixer is a good first step and a consistent way to wrap code in the "return a Promise" paradigm. Do we also want to document using [`.resolves`](https://jestjs.io/docs/en/expect#resolves), [`.rejects`](https://jestjs.io/docs/en/expect#rejects) or returning a Promise / await statement as other suggested alternatives?

Reply:
The thinking is that returning a promise is a safe change, as it will always work (I think?). Messing with the body of the tests are both way harder and more brittle.

Reply:
oh, I now understood that you meant the docs... I thought you talked about the fixer, sorry!

Yeah, we should definitely include some prose about using at least `async-wait` over returning the promise (async functions returns promises implicitly, so it's almost always a better idea anyways). Mentioning `resolves` and `rejects` (especially `rejects` as `resolves` i better handled by `await`) is probably a good idea as well.

Would you mind sending a PR for it? 🙏

Reply:
@SimenB rejects is very useful to avoid using logic in try/catch blocks (like `.toThrow` but for async functions), so I would keep it (and `resolves` for consistency) unless we come up with an equally good alternative.

Reply:
Yeah, for sure.

What I want is same as ava (which returns the error from `t.throws`, but I'm not sure if that makes sense for `expect`

Reply:
I just saw that ava has `throwsAsync` for that :D

Reply:
Ah, new in v1, it was just `throws` before. But yeah, I like that pattern. Doesn't fit into `expect`, though